WalletWasabi.Fluent/ViewModels/Wallets/Send/PrivacySuggestionsFlyoutViewModel.cs

```diff
@@ -36,45 +44,68 @@ public PrivacySuggestionsFlyoutViewModel()
+		using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(timeoutCts.Token, singleRunCts.Token, cancellationToken);
 
-		var loadingRing = new LoadingSuggestionViewModel();
-		Suggestions.Add(loadingRing);
+		using (await _asyncLock.LockAsync(CancellationToken.None))
+		{
+			try
+			{
+				Suggestions.Clear();
+				SelectedSuggestion = null;
 
-		var hasChange = transaction.InnerWalletOutputs.Any(x => x.ScriptPubKey != destination.ScriptPubKey);
+				var loadingRing = new LoadingSuggestionViewModel();
+				Suggestions.Add(loadingRing);
 
-		if (hasChange && !isFixedAmount && !info.IsPayJoin)
-		{
-			// Exchange rate can change substantially during computation itself.
-			// Reporting up-to-date exchange rates would just confuse users.
-			decimal usdExchangeRate = wallet.Synchronizer.UsdExchangeRate;
+				var hasChange = transaction.InnerWalletOutputs.Any(x => x.ScriptPubKey != destination.ScriptPubKey);
+
+				if (hasChange && !isFixedAmount && !info.IsPayJoin)
+				{
+					// Exchange rate can change substantially during computation itself.
+					// Reporting up-to-date exchange rates would just confuse users.
+					decimal usdExchangeRate = wallet.Synchronizer.UsdExchangeRate;
+
+					// Only allow to create 1 more input with BnB. This accounts for the change created.
+					int maxInputCount = transaction.SpentCoins.Count() + 1;
 
-			// Only allow to create 1 more input with BnB. This accounts for the change created.
-			int maxInputCount = transaction.SpentCoins.Count() + 1;
+					var pockets = wallet.GetPockets();
+					var spentCoins = transaction.SpentCoins;
+					var usedPockets = pockets.Where(x => x.Coins.Any(coin => spentCoins.Contains(coin)));
+					var coinsToUse = usedPockets.SelectMany(x => x.Coins).ToImmutableArray();
 
-			var pockets = wallet.GetPockets();
-			var spentCoins = transaction.SpentCoins;
-			var usedPockets = pockets.Where(x => x.Coins.Any(coin => spentCoins.Contains(coin)));
-			var coinsToUse = usedPockets.SelectMany(x => x.Coins).ToImmutableArray();
+					IAsyncEnumerable<ChangeAvoidanceSuggestionViewModel> suggestions =
+						ChangeAvoidanceSuggestionViewModel.GenerateSuggestionsAsync(info, destination, wallet, coinsToUse, maxInputCount, usdExchangeRate, linkedCts.Token);
 
-			IAsyncEnumerable<ChangeAvoidanceSuggestionViewModel> suggestions =
-				ChangeAvoidanceSuggestionViewModel.GenerateSuggestionsAsync(info, destination, wallet, coinsToUse, maxInputCount, usdExchangeRate, linkedCts.Token);
+					await foreach (var suggestion in suggestions)
+					{
+						Suggestions.Insert(Suggestions.Count - 1, suggestion);
+					}
+				}
 
-			await foreach (var suggestion in suggestions)
+				Suggestions.Remove(loadingRing);
+			}
+			catch (OperationCanceledException)
 			{
-				Suggestions.Insert(Suggestions.Count - 1, suggestion);
+				Logger.LogTrace("Operation was cancelled.");
+			}
+			finally
+			{
+				lock (_lock)
+				{
+					_suggestionCancellationTokenSource = null;
```

Comment:
Feels weird that we assign `null` instead of disposing it. 
But it probably makes zero difference, since the GC will collect it anyways.